unit test flake: pkg/master is flaky

This has just happened:

--- FAIL: TestGetServersToValidate (0.20s)
	<autogenerated>:31: 
                        
	Error Trace:	master_test.go:131
		
	Error:		Not equal: 3 (expected)
			        != 4 (actual)
		
	Messages:	unexpected server list: map[string]apiserver.Server{"controller-manager":apiserver.Server{Addr:"127.0.0.1", Port:10252, Path:"/healthz", EnableHTTPS:false, Validate:(apiserver.ValidatorFn)(nil)}, "scheduler":apiserver.Server{Addr:"127.0.0.1", Port:10251, Path:"/healthz", EnableHTTPS:false, Validate:(apiserver.ValidatorFn)(nil)}, "etcd-0":apiserver.Server{Addr:"", Port:4001, Path:"/health", EnableHTTPS:false, Validate:(apiserver.ValidatorFn)(0xb915a0)}, "etcd-1":apiserver.Server{Addr:"127.0.0.1", Port:49244, Path:"/health", EnableHTTPS:false, Validate:(apiserver.ValidatorFn)(0xb915a0)}}


https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/bd58b05f3eaea3d3b21df7d80f8b0a0c6c2ec87e/kubernetes-pull-test-unit-integration/7420/artifacts/?debugUI=DEVELOPERS

--- a/pkg/storage/etcd/etcd_helper.go
+++ b/pkg/storage/etcd/etcd_helper.go
@@ -130,10 +130,7 @@ func (h *etcdHelper) Backends(ctx context.Context) []string {
 		glog.Errorf("Error obtaining etcd members list: %q", err)
 		return nil
 	}
-	if 0 == len(members) {
-		return nil
-	}
-	mlist := []string{""}
+	mlist := []string{}
 	for _, member := range members {
 		mlist = append(mlist, member.ClientURLs...)
 	}

--- a/pkg/storage/etcd/testing/utils.go
+++ b/pkg/storage/etcd/testing/utils.go
@@ -32,6 +32,8 @@ import (
 	"github.com/coreos/etcd/pkg/transport"
 	"github.com/coreos/etcd/pkg/types"
 	"github.com/coreos/etcd/rafthttp"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
 )
 
 // EtcdTestServer encapsulates the datastructures needed to start local instance for testing
@@ -124,6 +126,22 @@ func (m *EtcdTestServer) launch(t *testing.T) error {
 	return nil
 }
 
+// waitForEtcd wait until etcd is propagated correctly
+func (m *EtcdTestServer) waitUntilUp() error {
+	membersAPI := etcd.NewMembersAPI(m.Client)
+	for start := time.Now(); time.Since(start) < 5*time.Second; time.Sleep(10 * time.Millisecond) {
+		members, err := membersAPI.List(context.TODO())
+		if err != nil {
+			glog.Errorf("Error when getting etcd cluster members")
+			continue
+		}
+		if len(members) == 1 && len(members[0].ClientURLs) > 0 {
+			return nil
+		}
+	}
+	return fmt.Errorf("timeout on waiting for etcd cluster")
+}
+
 // Terminate will shutdown the running etcd server
 func (m *EtcdTestServer) Terminate(t *testing.T) {
 	m.Client = nil
@@ -150,8 +168,13 @@ func NewEtcdTestClientServer(t *testing.T) *EtcdTestServer {
 	}
 	server.Client, err = etcd.New(cfg)
 	if err != nil {
-		t.Errorf("Unexpected Error in NewEtcdTestClientServer (%v)", err)
-		defer server.Terminate(t)
+		t.Errorf("Unexpected error in NewEtcdTestClientServer (%v)", err)
+		server.Terminate(t)
+		return nil
+	}
+	if err := server.waitUntilUp(); err != nil {
+		t.Errorf("Unexpected error in waitUntilUp (%v)", err)
+		server.Terminate(t)
 		return nil
 	}
 	return server

